homeassistant/components/ssdp/__init__.py

```diff
@@ -85,6 +86,29 @@
 _LOGGER = logging.getLogger(__name__)
 
 
+_SsdpServiceInfoBase = TypedDict(
+    "_SsdpServiceInfoBase",
+    {
+        "x-homeassistant-matching-domains": set,
+    },
+    total=True,
+)
+
+
+class SsdpServiceInfo(_SsdpServiceInfoBase, total=False):
+    """Prepared info from ssdp entries."""
+
+    # Attributes for accessing info from SSDP response
+    ssdp_location: str
+
+    # Attributes for accessing info from retrieved UPnP device description
+    friendlyName: str
+    manufacturer: str
+    serialNumber: str
+    UDN: str
+    presentationURL: str
```

Comment:
Should we add extra (non-required) attributes, or should we keep them for a follow-up PR?

```suggestion
    deviceType: str
    friendlyName: str
    manufacturer: str
    manufacturerURL: str
    modelDescription: str
    modelName: str
    modelNumber: str
    modelURL: str
    serialNumber: str
    serviceList: str
    UDN: str
    UPC: str
    presentationURL: str
```

Reply:
Perhaps the additional fields coming from the UPnP service description can be added in a subclass, e.g.
```py
class UpnpDescription(TypedDict):
    deviceType: str
    friendlyName: str
    # etc, for all fields all fields originally named by an ATTR_UPNP_ const at the top of this file.

class SsdpUpnpInfo(SsdpServiceInfo, UpnpDescription):
    pass
```

Then instances of `UpnpDescription` would be returned by `ssdp.Scanner._async_get_description_dict` and combined with a `SsdpServiceInfo` in `ssdp.Scanner._async_headers_to_discovery_info` to give `SsdpUpnpInfo`, which is what would be passed into `ConfigFlow.async_step_ssdp`.

Adding a bit of complication, some of those suggested fields aren't actually `str`. E.g. `serviceList` is a `list[list[dict[str,str]]]`. I can grab an example discovery info created by ssdp, if that would help.